src/octoprint/plugins/gcodeviewer/static/js/viewer/reader.js

```diff
@@ -117,41 +125,39 @@ GCODE.gCodeReader = (function () {
         return {layer: layer, cmd: cmd};
     };
 
-    var sanitizeModel = function (m) {
+    var cleanModel = function (m) {
         if (!m) return [];
-        return _.filter(m, function (layer) {
-            return !!layer;
-        });
-    };
 
-    var purgeEmptyLayers = function (m) {
-        return _.filter(m, function (layer) {
-            return (
-                !!layer &&
-                layer.length > 0 &&
-                _.find(layer, function (cmd) {
-                    return cmd && cmd.extrude;
-                }) !== undefined
-            );
-        });
+        var result = [];
+        rendererEmptyLayers = [];
+        rendererPercentageByLayer = [];
+        for (var i = 0; i < m.length; i++) {
+            if (!m[i]) continue;
+            if (gCodeOptions["purgeEmptyLayers"] && emptyLayers[i]) continue;
```

Comment:
`m[i]` skips an entry where the map has no elements for this index. No point in copying an empty index. Perhaps this cannot happen but I am unsure.
The next line skips a layer where there might be commands in the layer but there is no extrusion for that layer.

Frankly, I don't even know why this whole "purgeEmptyLayers" option even exists and I certainly don't understand why anyone would want to bother the end-user with this. The whole gcodeviewer code would be a lot simpler if this option didn't exist, regardless of which value you would chose to hardcode. We could just purge these layers during parsing or we would have an empty layer on screen with only a few move-without-extrucding commands (green lines).

It would also make this whole rebuilding of the maps unnecessary and all the work could happen in the worker. This only happens in the reader so that when you turn the option on or off, the data passed from the worker can be reprocessed to apply or unapply the option without downloading and parsing the whole gcode file again.

This is essentially the difference between  `rendererPercentageByLayer` and `percentageByLayer`. The `rendererPercentageByLayer` version has all the empty indexes removed and all the layers without extrusion if the "purgeEmptyLayers" option is set.  It is the cleaned data passed to the renderer.

Reply:
@cp2004 found out this option got introduced in this issue #1973

It wasn't really clear from the issue why this needed to be configurable. My best guess was from the issue, it was done in case it would have nasty side effects. Since nobody seemed to remember that, it seems it does not.

Anyway, your call.